specifyweb/frontend/js_src/lib/components/TreeView/AddRank.tsx

```diff
@@ -0,0 +1,158 @@
+  treeDefinitionItems,
+}: {
+  readonly tableName: SCHEMA['tableName'];
+  readonly treeDefinitionItems: RA<
+    SerializedResource<FilterTablesByEndsWith<'TreeDefItem'>>
+  >;
+}): JSX.Element {
+  const loading = React.useContext(LoadingContext);
+
+  const [isAddingRank, setIsAddingRank] = React.useState(false);
+  const [isAddingParentRank, setIsAddingParentRank] = React.useState(false);
+
+  const [parentRank, setParentRank] = React.useState('');
+  const treeId = strictIdFromUrl(treeDefinitionItems[0].treeDef);
+
+  const treeResource = React.useMemo(
+    () =>
+      tableName === 'Geography'
+        ? new tables.GeographyTreeDefItem.Resource({
+            _tableName: 'GeographyTreeDefItem',
+          })
+        : tableName === 'Taxon'
+        ? new tables.TaxonTreeDefItem.Resource({
+            _tableName: 'TaxonTreeDefItem',
+          })
+        : tableName === 'LithoStrat'
+        ? new tables.LithoStratTreeDefItem.Resource({
+            _tableName: 'LithoStratTreeDefItem',
+          })
+        : new tables.StorageTreeDefItem.Resource({
+            _tableName: 'StorageTreeDefItem',
+          }),
+    [tableName]
+  );
+
+  function addRank(): void {
+    const url = `/api/specify_tree/${tableName.toLowerCase()}/add_tree_rank/`;
+    loading(
+      ping(url, {
+        method: 'POST',
+        body: {
+          newRankName: (
+            treeResource as SpecifyResource<GeographyTreeDefItem>
+          ).get('name'),
+          parentRankName: parentRank,
+          treeID: treeId,
+          newRankTitle: (
+            treeResource as SpecifyResource<GeographyTreeDefItem>
+          ).get('title'),
+          remarks: (treeResource as SpecifyResource<GeographyTreeDefItem>).get(
+            'remarks'
+          ),
+          textAfter: (
+            treeResource as SpecifyResource<GeographyTreeDefItem>
+          ).get('textAfter'),
+          textBefore: (
+            treeResource as SpecifyResource<GeographyTreeDefItem>
+          ).get('textBefore'),
+          isEnforced: (
+            treeResource as SpecifyResource<GeographyTreeDefItem>
+          ).get('isEnforced'),
+          isInFullName: (
+            treeResource as SpecifyResource<GeographyTreeDefItem>
+          ).get('isInFullName'),
+          fullNameSeparator: (
+            treeResource as SpecifyResource<GeographyTreeDefItem>
+          ).get('fullNameSeparator'),
+        },
+      }).then(() => globalThis.location.reload())
+    );
+    setIsAddingRank(false);
+  }
+
+  return (
+    <>
+      <Button.Icon
+        icon="plus"
+        title={treeText.addNewRank()}
+        onClick={() => setIsAddingParentRank(true)}
+      />
+      {isAddingParentRank && (
+        <Dialog
+          buttons={
+            <>
+              <Button.DialogClose>{commonText.cancel()}</Button.DialogClose>
+              <Button.Save
+                onClick={() => {
+                  setIsAddingRank(true);
+                  setIsAddingParentRank(false);
```

Comment:
since these two states are mutually exclusive, why not create a state like this instead?
```ts
const [state, setState] = React.useState<'initial' | 'parent' | 'edit'>('initial');
```